In [68]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
df = pd. read_csv("/content/drive/MyDrive/Dataset/student.csv")
df.head(3)


Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB


In [70]:
df.describe()

Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000

In [71]:
X = df.drop(columns=["Writing"]).values
y = df["Writing"].values
# Number of features
d = X.shape[1]
# Initialize weights
W = np.zeros(d)
print(d)
print(W)
print(W.shape)  # (2,1)



# Perform a train - test split from scratch using a 80% − 20% ratio
# Define a function for train-test split from scratch
def train_test_split_scratch(X, y, test_size=0.2, random_seed=42):

  np.random.seed(random_seed)
  indices = np.arange(X.shape[0])
  #print(indices)
  np.random.shuffle(indices) # Shuffle the indices
  test_split_size = int(len(X) * test_size)
  test_indices = indices[:test_split_size]
  train_indices = indices[test_split_size:]
  X_train, X_test = X[train_indices], X[test_indices]
  y_train, y_test = y[train_indices], y[test_indices]
  return X_train, X_test, y_train, y_test
# Perform the train-test split
X_train, X_test, y_train, y_test = train_test_split_scratch(X, y, test_size=0.2)
# Output shapes to verify
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)



2
[0. 0.]
(2,)
Shape of X_train: (800, 2)
Shape of X_test: (200, 2)
Shape of y_train: (800,)
Shape of y_test: (200,)


In [72]:
def cost_function(X, y, W):
    """
    Calculates Mean Squared Error cost
    """
    n = len(y)
    Y_pred = np.dot(X, W)
    cost = (1 / (2 * n)) * np.sum((Y_pred - y) ** 2)
    return cost

costs=cost_function(X,y,W)
print(costs)

2470.11


In [73]:
# #testing cost function
# X_test = np.array([[1, 2], [3, 4], [5, 6]])
# Y_test = np.array([3, 7, 11])
# W_test = np.array([1, 1])
# cost = cost_function(X_test, Y_test, W_test)
# if cost == 0:
#   print("Proceed Further")
# else:
#   print("something went wrong: Reimplement a cost function")
# print("Cost function output:", cost_function(X_test, Y_test, W_test))

In [74]:
def gradient_descent(X, y, W, alpha, iterations):
    """
    Perform gradient descent to optimize linear regression parameters.

    Parameters:
    X : numpy.ndarray (m, d)  -> Feature matrix
    y : numpy.ndarray (m,)    -> Target vector
    W : numpy.ndarray (d,)    -> Initial weights
    alpha : float             -> Learning rate
    iterations : int          -> Number of iterations

    Returns:
    W : numpy.ndarray (d,)    -> Optimized weights
    cost_history : list       -> Cost at each iteration
    """

    m = len(y)                         # number of training samples
    cost_history = []

    for i in range(iterations):

        # Step 1: Prediction (Hypothesis)
        y_pred = np.dot(X, W)

        # Step 2: Error (loss)
        error = y_pred - y

        # Step 3: Gradient computation
        dw = (1 / m) * np.dot(X.T, error)

        # Step 4: Update weights
        W = W - alpha * dw
        W_update = W

        # Step 5: Compute cost
        cost = cost_function(X, y, W)
        cost_history.append(cost)

    return W_update, cost_history


In [75]:
# # Generate random test data
# np.random.seed(0) # For reproducibility
# X = np.random.rand(100, 3) # 100 samples, 3 features
# Y = np.random.rand(100)
# W = np.random.rand(3) # Initial guess for parameters
# # Set hyperparameters
# alpha = 0.01
# iterations = 1000
# # Test the gradient_descent function
# final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)
# # Print the final parameters and cost history
# print("Final Parameters:", final_params)
# print("Cost History:", cost_history)



In [76]:
def rmse(Y, Y_pred):
    """
    Calculates Root Mean Squared Error (RMSE)

    Parameters:
    Y : array-like -> Actual target values
    Y_pred : array-like -> Predicted values

    Returns:
    rmse : float -> Root Mean Squared Error
    """
    n = len(Y)
    mse = (1 / n) * np.sum((Y - Y_pred)**2)
    return np.sqrt(mse)



In [77]:
def r2(Y, Y_pred):
    """
    Calculates R-squared (Coefficient of Determination)

    Parameters:
    Y : array-like -> Actual target values
    Y_pred : array-like -> Predicted values

    Returns:
    r_squared : float -> R² score
    """
    mean_y = np.mean(Y)
    ss_tot = np.sum((Y - mean_y)**2)      # Total sum of squares
    ss_res = np.sum((Y - Y_pred)**2)      # Residual sum of squares
    r_squared = 1 - (ss_res / ss_tot)
    return r_squared


In [78]:
def main():

    # Step 1: Load the dataset
    # data = pd.read_csv("/content/drive/MyDrive/Dataset/student.csv")


    # Step 2: Split data into Features (X) and Target (Y)
    # X = data[['Math', 'Reading']].values   # Features: Math & Reading marks
    # Y = data['Writing'].values             # Target: Writing marks


    # Step 3: Train-Test Split (80% train, 20% test)
    X_train, X_test, Y_train, Y_test = train_test_split_scratch(X, y, test_size=0.2)


    # Step 4: Initialize weights, learning rate, iterations
    # W = np.zeros(X_train.shape[1])
    alpha = 0.00001
    iterations = 1000  # Number of gradient descent iterations


    # Step 5: Perform Gradient Descent
    W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)


    # Step 6: Make predictions on the test set
    Y_pred = np.dot(X_test, W_optimal)

    # Step 7: Evaluate the model
    model_rmse = rmse(Y_test, Y_pred)
    model_r2   = r2(Y_test, Y_pred)


    # Step 8: Output the results
    print("Final Weights:", W_optimal)
    print("Cost History (First 10 iterations):", cost_history[:10])
    print("RMSE on Test Set:", model_rmse)
    print("R-Squared on Test Set:", model_r2)

# Execute the main function
if __name__ == "__main__":
    main()


Final Weights: [0.34811659 0.64614558]
Cost History (First 10 iterations): [np.float64(2013.165570783755), np.float64(1640.286832599692), np.float64(1337.0619994901588), np.float64(1090.4794892850578), np.float64(889.9583270083234), np.float64(726.8940993009545), np.float64(594.2897260808594), np.float64(486.4552052951635), np.float64(398.7634463599484), np.float64(327.4517147324688)]
RMSE on Test Set: 5.2798239764188635
R-Squared on Test Set: 0.8886354462786421


Finding:
Our linear regression model performed well on the dataset predicting Writing marks from Math and Reading marks.

The model achieved an R² of 0.889 and RMSE of 5.28 on the test set.

The model is neither overfitting nor underfitting; training and test performance are consistent.

Learning rate experiments showed that smaller rates slow down convergence, while too high rates can cause divergence. The current learning rate (0.00001) provides stable and effective training.